### ISSUE TITLE
Feature request: Transpose multiple selections

### ISSUE THREAD

**Issue by dkeenan7** [notepad-plus-plus/notepad-plus-plus#6226]
When multi-editing (Ctrl+select) is enabled, and there are 2 selections, then Ctrl+T (Transpose) should swap the contents of those 2 selections. This should not affect the clipboard, and it should not affect which selection-position is the main selection. If there is only 1 selection, Ctrl+T should do as it does now (transpose lines). If there are 3 or more selections it should rotate their contents. 

The same should apply to the rows of a rectangular or column selection (Alt-select) irrespective of whether multi-editing is enabled.

The Edit menu should contain an item "Transpose    Ctrl+T". It might change from "Transpose Lines" to "Transpose Selections" to "Transpose Rows" depending on the nature of the selection(s).

This feature exists in Sublime Text:
https://stackoverflow.com/questions/16801313/swapping-text-selections-in-sublime-text-2
and in Visual Studio (via a plug-in):
https://marketplace.visualstudio.com/items?itemName=SSDiver2112.SwapSelection

**Comment by dinkumoil** [comment]
Since the transpose functionality is handled by Scintilla (the edit component Notepad++ uses internally), you should file a feature request at the [Scintilla project site](https://sourceforge.net/p/scintilla/feature-requests/). For this reason you should close the issue here as well.

If you need this functionality very quickly, go to the [Notepad++ user forum](https://community.notepad-plus-plus.org/) and ask for support to implement it with the help of a scripting plugin (e.g. _PythonScript_ plugin, _LuaScript_ plugin, _ActiveX_ plugin).

**Comment by dkeenan7** [comment]
Thanks dinkumoil. I don't need it quickly. I've waited 35 years. A few more months won't matter. ;-) I'd prefer it was done properly. I have now filed this Scintilla feature request:
https://sourceforge.net/p/scintilla/feature-requests/1316/
I'd prefer not to close it here until there's some response from the Scintilla devs. And in any case, the addition of "Transpose Ctrl+T" or "Transpose Lines Ctrl+T" to the Edit menu (even just as a form of documentation for the existing command) requires a change to Notepad++, not Scintilla.

And I expect the full requested functionality can still be implemented in Notepad++ if the Scintilla devs decline.

**Comment by sasumner** [comment]
@dkeenan7 said:

>If there are 3 or more selections it should rotate their contents.

How about providing a concrete example?

Also, how about a real use-case where this (3+ rotation) is valuable to do?  Just because other editors can do it doesn't really prove it is useful.

>The same should apply to the rows of a rectangular or column selection

I'll get an example of this started...what would you expect for the following?:

![image](https://user-images.githubusercontent.com/30118311/67008198-01b79280-f0b7-11e9-921f-2c4c644387cc.png)

Here's the data as text if you want to "play" with it:

```
   aaa
   bbbbbb
   cccccccc
   ddddddddddd
   eeeeeeeeeeeee
   fff
   gggggg
   hhhhhhh

```

**Comment by dkeenan7** [comment]
Hi sasumner. I take it that you find the usefulness of the 2 selection swap to be obvious enough. 

To be honest. I think the 3+ cases would rarely be useful. It just seems a shame not to make Ctrl+T do _something_ in those cases, and rotation seems the obvious generalisation of swapping 2 items. If you implement rotation, then you naturally get swapping when the selection count is 2. No special case. And the word "transpose" already has the meaning of two _or more_ things changing places.

But I can remember times when I have typed values into a table and I didn't notice that I missed a value until I had finished the whole table. With this feature, I could type the missing value at the end and then Alt+select and Ctr+T to rotate it into the correct position.

After one Ctrl+T, your example selection would be rotated down one row, with the bottom item coming to the top, and so would look like the following. 
```
   aahhh
   bba  bb
   ccbbbccc
   ddcccdddddd
   eedddeeeeeeee
   ffeee
   ggf  g
   hhggghh
```
It would have the same hiliting as your original image. I'm assuming it was selected from top to bottom, if that matters. 

Because there are 8 selections in this example, typing Ctrl+T 7 times would be equivalent to rotating it upwards once, and Ctrl+T 8 times would put it back as it was.

The Scintilla documentation says that a rectangular selection is treated as multiple selections, one for each line. So there is no special case for rectangular versus discontiguous selections either.
In pseudocode, the algorithm would be:
```
n = numSelections;
temp = selection[n];
for i = n downto 2
    selection[i] = selection[i-1];
selection[1] = temp;
```

**Comment by sasumner** [comment]
@dkeenan7 When you showed what would happen with my "aaa...hhh" column block of data, note that when you rotated the "a" lines  and the "f" lines, you actually changed the data involved -- you added in two spaces (on each line of those 2 lines) that weren't there originally.  Is that always a reasonable thing to do?

So if one continued to repeatedly rotate, soon the "f" line would rotate into the "a" line.  Since both the "a" and the "f" lines originally contained just a single letter, when "f" rotates into "a", would it retain the two extra spaces added when rotating into the earlier lines?  Or would the algorithm know that no additional spaces on the "f" are necessary to fulfill a rotation into the "a" line (because of what the "a" line looked like originally)?

Thus, do you think it should look like this when done with the above rotation sequence?:

![image](https://user-images.githubusercontent.com/30118311/67405139-58730f80-f582-11e9-967a-bacb4c79c2b1.png)

Or like this?: (note the only difference is in the first line)

![image](https://user-images.githubusercontent.com/30118311/67405320-9708ca00-f582-11e9-9d13-ff7e8a2e1b6c.png)

**Comment by dkeenan7** [comment]
@sasumner That's a good point, and a good question. This is a question about "virtual space" and when should virtual spaces become real spaces. This is definitely a question for Scintilla, not Notepad++. I'm happy to accept whatever the Scintilla text engine does when you simply reassign the multiple selections as per my pseudocode above. It should not keep track of what the lines looked like originally.

I would prefer your second result above, and I note that it is not necessary to keep track of what the lines looked like originally, in order to achieve it. It is only necessary to postpone the conversion of virtual spaces into real spaces until the time that the selection rectangle ceases to exist, or changes, e.g. due to clicking with the mouse or using the arrow keys. While the selection rectangle remains unchanged, I would expect it to look like this.

![image](https://user-images.githubusercontent.com/13585061/67444595-1cf14780-f64d-11e9-8ad1-2018838e703e.png)

Then when the selection rectangle goes away, any virtual spaces that are followed by characters other that CRLF on the same line, should be turned into real spaces, giving your second result above.

**Comment by sasumner** [comment]
@dkeenan7 said:

>This is definitely a question for Scintilla, not Notepad++...I'm happy to accept whatever the Scintilla text engine does when you simply reassign the multiple selections 

I think it is a question for whomever is spec'ing out the feature, since it doesn't currently exist. :)

A "problem" comes in after the first shift.  The "f" line rotates into the "g" line and results in `ggf  g`.  Now the selected part contains `f space space`.  Now we rotate some more and a bit later the `f space space` ends up in the "a" line.  We now have no idea (without "keeping track" from the beginning) if those two spaces are real (were actually there in the original "f" line as spaces, not virtual space) or not.  So the logic has no idea if they should be kept or eliminated.  Admittedly, it is a "finer" point...

**Comment by dkeenan7** [comment]
What I'm suggesting, to avoid that problem, is that after the first shift, the selected part would not contain `f space space`, but only `f`, even though it would be displayed with a gap of the same width as two spaces, between the f and the following g, in the same way that there is a gap of two spaces between the f and the following CRLF when the selection is first made. So all virtual spaces would remain virtual spaces, through any number of rotations. Only when you click elsewhere, would the decision have to be made, about which virtual spaces should become real spaces. And that decision only depends on whether there are non CRLF characters following it on the same line.

My understanding is that a rectangular selection is represented as a list of selections, one for each row. But this is supplemented with information about the rectangle. The individual row selections may have any length from zero up to the width of the rectangle. But on-screen, any following characters (such as CRLF) are moved to the right as if every row consisted of rectangle-width characters.

**Comment by dinkumoil** [comment]
@dkeenan7  said:

> What I'm suggesting, to avoid that problem, is that after the first shift, the selected part would not contain f space space, but only f, even though it would be displayed with a gap of the same width as two spaces, between the f and the following g

I'm not sure if Scintilla supports virtual spaces in the middle of a line, i.e. between other characters. If not your approch is impossible.

**BTW:** Because I got interested in that task I already wrote some Lua code to solve it, thus it requires the _Lua script_ plugin to be installed. It installs two menu entries to rotate up- and downwards. These can be assigned to keyboard shortcuts by the _Shortcut Mapper_ dialog.

The code can handle whole-line and rectangular selections, empty selections (one only pressed CTRL and clicked with the mouse), can rotate down- and upwards and wraps every transpose operation in a transaction to be able to undo the operation as a whole. The main selection remains in that line number where it was when the operation has been started. The order in which the selections where made doesn't matter, the selections are sorted by their document position and get processed in that order.

Here is how the code handles a rectangular selection that includes virtual spaces:

Initial state:

![grafik](https://user-images.githubusercontent.com/24699812/67492735-2f807680-f677-11e9-9187-9fc2ec32d70d.png)

After rotating downwards:

![grafik](https://user-images.githubusercontent.com/24699812/67492851-576fda00-f677-11e9-91a4-761780e417fe.png)

After rotating upwards again:

![grafik](https://user-images.githubusercontent.com/24699812/67492899-65255f80-f677-11e9-9aa6-32b17b2fe427.png)

If you are interested in the code come to the [community forum](https://community.notepad-plus-plus.org/) and open a topic.

**Comment by sasumner** [comment]
@dinkumoil 

Ah, I had not been thinking in terms of the originally-rectangular selection becoming "unrectangular" due to the special circumstances of my original example.  But I like the idea as it clears up the "problems".  :-)  This is a great example of what "talking through" ideas here does.

>If you are interested in the code come to the community forum and open a topic.

Why not post code directly here if you want to share?

**Comment by dinkumoil** [comment]
@sasumner said:

> Why not post code directly here if you want to share?

Because here most of the people will never find the code. Opening a topic in the community forum with an appropriate title to reflect the problem will prevent that.

**Comment by sasumner** [comment]
@dinkumoil I was thinking more in terms of possibly aiding a developer that might add this feature in the future, in which case IMO opinion it would belong here; at least that's typically what I do with such scripts.  But regardless, do what you will. :)

**Comment by dinkumoil** [comment]
@sasumner OK, here is the code. It has to be added to the startup script of the _Lua Script_ plugin (see `(menu) Plugins -> LuaScript -> Edit Startup Script`). After restarting Notepad++ the two menu entries `Transpose Selections Up` and `Transpose Selections Down` will be available under `(menu) Plugins -> LuaScript` and can be assigned to keyboard shortcuts.

```Lua
-- =============================================================================
-- Add menu entries to transpose multiple selections by moving them up or down
-- =============================================================================

-- Constants for transpose directions
local TRANSPOSE_SEL_DIR_UP   = 1
local TRANSPOSE_SEL_DIR_DOWN = 2


-- Get text of selection with index idx
local function getSelectionNText(idx)
  local retVal = editor:textrange(editor.SelectionNStart[idx], editor.SelectionNEnd[idx])
  if not retVal then retVal = "" end

  return retVal
end


-- Set text of selection with index idx and readjust its span
local function setSelectionNText(idx, str)
  editor:SetTargetRange(editor.SelectionNStart[idx], editor.SelectionNEnd[idx])
  editor:ReplaceTarget(str)

  editor.SelectionNCaret[idx] = editor.SelectionNAnchor[idx] + #str
end


-- Use Insertion Sort to create a sorted list of selection indices
-- Using table.sort is avoided since it requires usage of 1 based indices
local function sortSelections(selCnt)
  local selPos, idxDst
  local selections = {}

  -- Create initial list
  for idx = 0, selCnt - 1 do
    selections[idx] = idx
  end

  -- Sort list
  for idxSrc = 1, selCnt - 1 do
    selPos = editor.SelectionNStart[idxSrc]
    idxDst = idxSrc

    while (idxDst > 0) and (editor.SelectionNStart[selections[idxDst - 1]] > selPos) do
      selections[idxDst] = selections[idxDst - 1]
      idxDst = idxDst - 1
    end

    selections[idxDst] = idxSrc
  end

  return selections
end


-- Transpose selections
-- If there is only 1 selection, the standard Line Transpose function is used
-- If there are more than 2 selections, they get rotated according to parameter
-- direction
local function transposeSelections(direction)
  local selCnt, bufStr
  local startIdx, endIdx, increment
  local srcIdx, dstIdx

  selCnt = editor.Selections

  -- Process special cases:
  --   0 selections -> do nothing
  --   1 selection  -> perform SCI_LINETRANSPOSE
  if selCnt < 1 then
    return
  elseif selCnt == 1 then
    npp.SendEditor(SCI_LINETRANSPOSE)
    return
  end

  -- Set loop parameters depending on transpose direction
  local params = {{0, selCnt - 1, 1}, {selCnt - 1, 0, -1}}
  startIdx, endIdx, increment = (function(p) return p[1], p[2], p[3] end)(params[direction])

  -- Create list of selection indices sorted by
  -- the starting position of the related selection
  local selections = sortSelections(selCnt)

  -- Set start of undo sequence
  editor:BeginUndoAction()

  -- Perform transpose operation
  bufStr = getSelectionNText(selections[startIdx])
  dstIdx = startIdx

  while dstIdx ~= endIdx do
    srcIdx = (dstIdx + increment) % selCnt
    setSelectionNText(selections[dstIdx], getSelectionNText(selections[srcIdx]))
    dstIdx = dstIdx + increment
  end

  setSelectionNText(selections[dstIdx], bufStr)

  -- Set end of undo sequence
  editor:EndUndoAction()
end


-- Add menu entry for transposing lines upwards
npp.AddShortcut("Transpose Selections Up", "", function()
  transposeSelections(TRANSPOSE_SEL_DIR_UP)
end)


-- Add menu entry for transposing lines downwards
npp.AddShortcut("Transpose Selections Down", "", function()
  transposeSelections(TRANSPOSE_SEL_DIR_DOWN)
end)
```

**Comment by dkeenan7** [comment]
@dinkumoil Thanks so much for that script! I have opened a topic for it on the community forum as you suggested.
https://community.notepad-plus-plus.org/topic/18415/transpose-multiple-selections/1

**Comment by dkeenan7** [comment]
We now have a Lua script that demonstrates the desired behaviour when rotating rectangular selections containing virtual space. It also implements transposing two characters for an empty selection not at the start or end of a line.
See https://community.notepad-plus-plus.org/topic/18415/transpose-multiple-selections/11

**Comment by donho** [comment]
@dkeenan7 
The idea is nice, and could be useful in the case of only 2 selections. OTOH, it won't be bend to Ctrl+T.
Do you still think it's an interesting feature for you?

**Comment by dkeenan7** [comment]
Yes. I still think it is worth having as a built-in. I assume I could choose to bind it to Ctrl+T if I don't want to transpose lines. I hope you have seen what Neil Hodgson has done to facilitate it in Scintilla. https://sourceforge.net/p/scintilla/feature-requests/1316/
It would be good if it also transposed the two characters either side of an empty selection, for correcting typos.

**Comment by sasumner** [comment]
@dkeenan7 
Can you summarize the Scintilla efforts on this a bit here, since you are closest to it?  I had a look at the Scintilla feature request link and there was a lot there.  In my mind, the work on this is 99% on the Scintilla side, and 1% (expose the SCI_XXXXXX command(s)) on the Notepad++ side.

**Comment by dkeenan7** [comment]
I'm glad you asked me to do that @sasumner . It turns out I had mis-remembered how far Neil Hodgson had got with this. I thought he had written these APIs:
SCI_ROTATESELECTIONS
SCI_BACKROTATESELECTIONS
SCI_REVERSESELECTIONS
But what Neil has done is to implement some helper routines, and fix a bug, that will make it much easier for him to implement the above APIs. I have just posted to the Scintilla thread to encourage him to take those final steps.

So @donho I suggest you wait until Neil has implemented the above APIs, before proceeding with this in Notepad++. Thankyou.

**Comment by znsoooo** [comment]
@dkeenan7 @dinkumoil @sasumner Hi, I've implemented this function that can swap multiple selected area texts. It was written in C++. You might be interested, here is the project:

<https://github.com/znsoooo/npp-select-in-brackets/releases>
